flink-runtime/src/main/java/org/apache/flink/runtime/highavailability/FileSystemJobResultStore.java

```diff
@@ -0,0 +1,210 @@
+        }
+
+        boolean deleteOnCommit = config.get(JobResultStoreOptions.DELETE_ON_COMMIT);
+
+        return new FileSystemJobResultStore(basePath.getFileSystem(), basePath, deleteOnCommit);
+    }
+
+    /**
+     * Given a job ID, construct the path for a dirty entry corresponding to it in the job result
+     * store.
+     *
+     * @param jobId The job ID to construct a dirty entry path from.
+     * @return A path for a dirty entry for the given the Job ID.
+     */
+    private Path constructDirtyPath(JobID jobId) {
+        return new Path(this.basePath.getPath() + "/" + jobId.toString() + DIRTY_SUFFIX + ".json");
+    }
+
+    /**
+     * Given a job ID, construct the path for a clean entry corresponding to it in the job result
+     * store.
+     *
+     * @param jobId The job ID to construct a clean entry path from.
+     * @return A path for a clean entry for the given the Job ID.
+     */
+    private Path constructCleanPath(JobID jobId) {
+        return new Path(this.basePath.getPath() + "/" + jobId.toString() + ".json");
+    }
+
+    @Override
+    public void createDirtyResultInternal(JobResultEntry jobResultEntry) throws IOException {
+        final Path path = constructDirtyPath(jobResultEntry.getJobId());
+        OutputStream os = fileSystem.create(path, FileSystem.WriteMode.NO_OVERWRITE);
+        mapper.writeValue(os, new JsonJobResultEntry(jobResultEntry));
+    }
+
+    @Override
+    public void markResultAsCleanInternal(JobID jobId) throws IOException, NoSuchElementException {
+        Path dirtyPath = constructDirtyPath(jobId);
+        if (fileSystem.exists(dirtyPath)) {
+            if (deleteOnCommit) {
+                fileSystem.delete(dirtyPath, false);
+            } else {
+                fileSystem.rename(dirtyPath, constructCleanPath(jobId));
+            }
+        } else {
+            throw new NoSuchElementException(
+                    String.format(
+                            "Could not mark job %s as clean as it is not present in the job result store.",
+                            jobId));
+        }
+    }
+
+    @Override
+    public boolean hasDirtyJobResultEntryInternal(JobID jobId) throws IOException {
+        return fileSystem.exists(constructDirtyPath(jobId));
+    }
+
+    @Override
+    public boolean hasCleanJobResultEntryInternal(JobID jobId) throws IOException {
+        return fileSystem.exists(constructCleanPath(jobId));
+    }
+
+    @Override
+    public Set<JobResult> getDirtyResultsInternal() throws IOException {
+        final Set<JobResult> dirtyResults = new HashSet<>();
+        final FileStatus fs = fileSystem.getFileStatus(this.basePath);
+        if (fs.isDir()) {
+            FileStatus[] statuses = fileSystem.listStatus(this.basePath);
+            for (FileStatus s : statuses) {
+                if (!s.isDir()) {
+                    String path = s.getPath().getPath();
+                    if (path.contains(DIRTY_SUFFIX)) {
```

Comment:
```suggestion
                    final String filename = s.getPath().getName();
                    if (filename.contains(DIRTY_SUFFIX)) {
```
nit: I don't know - this `getPath().getPath()` looks weird. What about using the filename instead? That might be sufficient enough.